File: frontend/src/types/graph.ts
import { z } from 'zod'

export const nodeTypeSchema = z.enum(['Paper', 'Author', 'Topic', 'Subfield', 'Field', 'Domain', 'External'])
export type NodeType = z.infer<typeof nodeTypeSchema>

const nullableString = z.preprocess((value) => value ?? undefined, z.string().optional())
const nullableNumber = z.preprocess((value) => value ?? undefined, z.number().optional())
const nullableBoolean = z.preprocess((value) => value ?? undefined, z.boolean().optional())

export const graphNodeSchema = z.object({
  id: z.string(),
  label: z.string(),
  type: nodeTypeSchema,
  val: nullableNumber,
  paperCount: nullableNumber,
  year: nullableNumber,
  citationCount: nullableNumber,
  read: nullableBoolean,
  unlocked: nullableBoolean,
  inScope: nullableBoolean,
  isAuthored: nullableBoolean,
  isReferenced: nullableBoolean,
  isCiting: nullableBoolean,
  x: nullableNumber,
  y: nullableNumber,
  vx: nullableNumber,
  vy: nullableNumber,
  fx: nullableNumber,
  fy: nullableNumber,
})
export type GraphNode = z.infer<typeof graphNodeSchema>

export const graphLinkSchema = z.object({
  source: z.union([z.string(), graphNodeSchema]),
  target: z.union([z.string(), graphNodeSchema]),
  type: z.string(),
})
export type GraphLink = z.infer<typeof graphLinkSchema>

export const graphDataSchema = z.object({
  nodes: z.array(graphNodeSchema),
  links: z.array(graphLinkSchema),
})
export type GraphData = z.infer<typeof graphDataSchema>

export const paperSchema = z.object({
  id: z.string(),
  title: nullableString,
  year: nullableNumber,
  abstract: nullableString,
  citationCount: nullableNumber,
})
export type Paper = z.infer<typeof paperSchema>

export const paperStatusSchema = z.enum(['to_read', 'reading', 'read', 'skipped'])
export type PaperStatus = z.infer<typeof paperStatusSchema>

export const authorSchema = z.object({
  id: z.string(),
  name: z.string(),
  institution: nullableString,
})
export type Author = z.infer<typeof authorSchema>

export const topicSchema = z.object({
  id: nullableString,
  name: z.string(),
  score: nullableNumber,
  paperCount: nullableNumber,
})

export const paperDetailSchema = paperSchema.extend({
  authors: z.array(authorSchema),
  topics: z.array(topicSchema),
  cites: z.array(paperSchema),
  cited_by: z.array(paperSchema),
  prerequisites: z.array(paperSchema),
  status: z.preprocess((value) => value ?? undefined, paperStatusSchema.optional()),
})
export type PaperDetail = z.infer<typeof paperDetailSchema>

export const learningPathItemSchema = z.object({
  id: z.string(),
  title: nullableString,
  year: nullableNumber,
  citationCount: nullableNumber,
  depth: z.number(),
  already_read: z.boolean(),
})

export const learningPathResultSchema = z.object({
  target: z.object({
    id: z.string(),
    title: nullableString,
  }),
  learning_path: z.array(learningPathItemSchema),
  papers_to_read: z.number(),
  ai_explanation: z.string().nullable(),
})
export type LearningPathResult = z.infer<typeof learningPathResultSchema>

export const topicResultSchema = z.object({
  id: z.string(),
  name: z.string(),
  works_count: nullableNumber,
  paperCount: nullableNumber,
})
export type TopicResult = z.infer<typeof topicResultSchema>

export const searchResultsSchema = z.object({
  papers: z.array(paperSchema),
  topics: z.array(topicResultSchema),
  authors: z.array(authorSchema.extend({
    paperCount: nullableNumber,
    totalCitations: nullableNumber,
  })).default([]),
})
export type SearchResults = z.infer<typeof searchResultsSchema>

export const paperInPlanRoleSchema = z.enum(['target', 'prerequisite', 'builds_on', 'topic_anchor'])

export const paperInPlanSchema = z.object({
  id: z.string(),
  title: nullableString,
  year: nullableNumber,
  citationCount: nullableNumber,
  abstract: nullableString,
  role: paperInPlanRoleSchema,
  depth: nullableNumber,
})
export type PaperInPlan = z.infer<typeof paperInPlanSchema>

export const lessonSectionSchema = z.object({
  heading: z.string(),
  content: z.string(),
})
export type LessonSection = z.infer<typeof lessonSectionSchema>

export const paperLessonSchema = z.object({
  paper_id: z.string(),
  lesson_title: z.string(),
  overview: z.string(),
  connection_to_previous: nullableString,
  why_now: z.string(),
  key_concepts: z.array(z.string()),
  lesson_sections: z.array(lessonSectionSchema),
  check_for_understanding: z.string(),
  knowledge_state_after: z.string(),
  grounded_in: z.string(),
})
export type PaperLesson = z.infer<typeof paperLessonSchema>

export const learnPlanResultSchema = z.object({
  artifact_id: nullableNumber,
  created_at: nullableString,
  target_title: z.string(),
  papers: z.array(paperInPlanSchema),
  plan: z.string(),
  curriculum: z.array(paperLessonSchema),
  total_papers: z.number(),
  provider: nullableString,
  model: nullableString,
  evidence_ids: z.array(z.string()).default([]),
})
export type LearnPlanResult = z.infer<typeof learnPlanResultSchema>

export const learnQuestionResponseSchema = z.object({
  answer: z.string(),
})
export type LearnQuestionResponse = z.infer<typeof learnQuestionResponseSchema>

export const successResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
})

export const paperStatusItemSchema = z.object({
  paper_id: z.string(),
  status: paperStatusSchema,
})
export const paperStatusListSchema = z.object({
  user_id: z.string(),
  statuses: z.array(paperStatusItemSchema),
})
export type PaperStatusList = z.infer<typeof paperStatusListSchema>

export const demoBootstrapSchema = z.object({
  user_id: z.string(),
  provider: z.string(),
  model: z.string(),
  configured: z.boolean(),
  featured_topics: z.array(z.string()),
  featured_professors: z.array(authorSchema.extend({
    paperCount: nullableNumber,
    totalCitations: nullableNumber,
  })),
})
export type DemoBootstrap = z.infer<typeof demoBootstrapSchema>

export const topicMasterySchema = z.object({
  topic: z.string(),
  provider: z.string(),
  model: z.string(),
  configured: z.boolean(),
  anchors: z.array(paperInPlanSchema),
  study_plan: z.string(),
  evidence_ids: z.array(z.string()),
})
export type TopicMastery = z.infer<typeof topicMasterySchema>

export const recommendationEvidenceSchema = z.object({
  paper_id: z.string(),
  relation: z.string(),
  note: nullableString,
})
export const recommendationItemSchema = z.object({
  paper: paperSchema,
  reason: z.string(),
  reason_code: z.string(),
  evidence: z.array(recommendationEvidenceSchema),
})
export const recommendationResponseSchema = z.object({
  artifact_id: nullableNumber,
  created_at: nullableString,
  user_id: z.string(),
  context: z.string(),
  provider: z.string(),
  model: z.string(),
  items: z.array(recommendationItemSchema),
})
export type RecommendationResponse = z.infer<typeof recommendationResponseSchema>

export const professorBriefSchema = z.object({
  artifact_id: nullableNumber,
  created_at: nullableString,
  professor: authorSchema,
  provider: z.string(),
  model: z.string(),
  configured: z.boolean(),
  research_brief: z.string(),
  industry_impact: z.string().default(''),
  build_on_research: z.string().default(''),
  approach_advice: z.string().default(''),
  future_directions: z.array(z.string()),
  evidence_ids: z.array(z.string()),
  top_papers: z.array(paperSchema),
  timeline: z.array(paperSchema),
  topic_names: z.array(z.string()),
  collaborators: z.array(z.object({
    id: z.string(),
    name: z.string(),
    institution: nullableString,
    shared_papers: nullableNumber,
  })),
  descendant_papers: z.array(paperSchema),
  referenced_papers: z.array(paperSchema).default([]),
  authored_paper_count: z.number().default(0),
  referenced_paper_count: z.number().default(0),
  citing_paper_count: z.number().default(0),
})
export type ProfessorBrief = z.infer<typeof professorBriefSchema>

export const generatedArtifactHistoryItemSchema = z.object({
  id: z.number(),
  artifact_type: z.string(),
  subject_type: z.string(),
  subject_id: z.string(),
  created_at: nullableString,
  provider: z.string(),
  model: z.string(),
})
export const generatedArtifactHistorySchema = z.object({
  items: z.array(generatedArtifactHistoryItemSchema),
})
export type GeneratedArtifactHistory = z.infer<typeof generatedArtifactHistorySchema>
export type GeneratedArtifactHistoryItem = z.infer<typeof generatedArtifactHistoryItemSchema>

export const generatedArtifactCatalogItemSchema = z.object({
  id: z.number(),
  artifact_type: z.string(),
  subject_type: z.string(),
  subject_id: z.string(),
  created_at: nullableString,
  provider: z.string(),
  model: z.string(),
  title: nullableString,
})
export const generatedArtifactCatalogSchema = z.object({
  items: z.array(generatedArtifactCatalogItemSchema),
})
export type GeneratedArtifactCatalog = z.infer<typeof generatedArtifactCatalogSchema>

File: backend/app/models/graph.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

NodeType = Literal["Paper", "Author", "Topic", "Subfield", "Field", "Domain", "External"]
ProfessorPaperRole = Literal["authored", "references", "citations"]
PlanPaperRole = Literal["target", "prerequisite", "builds_on", "topic_anchor"]
PaperStatusOption = Literal["to_read", "reading", "read", "skipped"]


class ApiModel(BaseModel):
    model_config = ConfigDict(extra="ignore")


class Author(ApiModel):
    id: str
    name: str
    institution: str | None = None


class Topic(ApiModel):
    id: str | None = None
    name: str
    score: float | None = None
    paperCount: int | None = None


class Paper(ApiModel):
    id: str
    title: str | None = None
    year: int | None = None
    abstract: str | None = None
    citationCount: int | None = None


class PaperDetail(Paper):
    authors: list[Author] = Field(default_factory=list)
    topics: list[Topic] = Field(default_factory=list)
    cites: list[Paper] = Field(default_factory=list)
    cited_by: list[Paper] = Field(default_factory=list)
    prerequisites: list[Paper] = Field(default_factory=list)
    status: PaperStatusOption | None = None


class AuthorDetail(Author):
    papers: list[Paper] = Field(default_factory=list)


class GraphNode(ApiModel):
    id: str
    label: str
    type: NodeType
    val: int | None = None
    paperCount: int | None = None
    citationCount: int | None = None
    year: int | None = None
    read: bool | None = None
    unlocked: bool | None = None
    inScope: bool | None = None
    isAuthored: bool | None = None
    isReferenced: bool | None = None
    isCiting: bool | None = None


class GraphLink(ApiModel):
    source: str
    target: str
    type: str


class GraphResponse(ApiModel):
    nodes: list[GraphNode]
    links: list[GraphLink]


class TopicResult(ApiModel):
    id: str
    name: str
    works_count: int | None = None
    paperCount: int | None = None


class SearchResponse(ApiModel):
    papers: list[Paper]
    topics: list[TopicResult]
    authors: list["ProfessorSearchResult"] = Field(default_factory=list)


class LearningPathItem(ApiModel):
    id: str
    title: str | None = None
    year: int | None = None
    citationCount: int | None = None
    depth: int
    already_read: bool = False


class LearningPathTarget(ApiModel):
    id: str
    title: str | None = None


class LearningPathResponse(ApiModel):
    target: LearningPathTarget
    learning_path: list[LearningPathItem]
    papers_to_read: int
    ai_explanation: str


class ReadBody(ApiModel):
    paper_id: str
    user_id: str | None = None


class ReadResponse(ApiModel):
    success: bool
    message: str


class PaperInPlan(ApiModel):
    id: str
    title: str | None = None
    year: int | None = None
    citationCount: int | None = None
    abstract: str | None = None
    role: PlanPaperRole
    depth: int | None = None


class LessonSection(ApiModel):
    heading: str
    content: str


class PaperLesson(ApiModel):
    paper_id: str
    lesson_title: str
    overview: str
    connection_to_previous: str | None = None
    why_now: str
    key_concepts: list[str] = Field(default_factory=list)
    lesson_sections: list[LessonSection] = Field(default_factory=list)
    check_for_understanding: str
    knowledge_state_after: str
    grounded_in: str


class LearnPlanResponse(ApiModel):
    artifact_id: int | None = None
    created_at: datetime | None = None
    target_title: str
    papers: list[PaperInPlan]
    plan: str
    curriculum: list[PaperLesson] = Field(default_factory=list)
    total_papers: int
    provider: str | None = None
    model: str | None = None
    evidence_ids: list[str] = Field(default_factory=list)


class LearnQuestionRequest(ApiModel):
    target_title: str
    paper: PaperInPlan
    lesson: PaperLesson
    learned_context: str = ""
    question: str


class LearnQuestionResponse(ApiModel):
    answer: str


class ExplainPathRequest(ApiModel):
    paper_ids: list[str] = Field(min_length=1)


class SummarizePaperRequest(ApiModel):
    paper_id: str


class FieldTrajectoryRequest(ApiModel):
    topic: str = Field(min_length=1)


class PaperStatusRequest(ApiModel):
    paper_id: str
    status: PaperStatusOption
    user_id: str | None = None
    note: str | None = None


class PaperStatusItem(ApiModel):
    paper_id: str
    status: PaperStatusOption


class PaperStatusListResponse(ApiModel):
    user_id: str
    statuses: list[PaperStatusItem] = Field(default_factory=list)


class ProfessorSearchResult(ApiModel):
    id: str
    name: str
    institution: str | None = None
    paperCount: int | None = None
    totalCitations: int | None = None


class RecommendationEvidence(ApiModel):
    paper_id: str
    relation: str
    note: str | None = None


class RecommendationItem(ApiModel):
    paper: Paper
    reason: str
    reason_code: str
    evidence: list[RecommendationEvidence] = Field(default_factory=list)


class RecommendationResponse(ApiModel):
    artifact_id: int | None = None
    created_at: datetime | None = None
    user_id: str
    context: str
    provider: str
    model: str
    items: list[RecommendationItem] = Field(default_factory=list)


class TopicMasteryResponse(ApiModel):
    topic: str
    provider: str
    model: str
    configured: bool
    anchors: list[PaperInPlan] = Field(default_factory=list)
    study_plan: str
    evidence_ids: list[str] = Field(default_factory=list)


class ProfessorBriefContext(ApiModel):
    professor: Author
    top_papers: list[Paper] = Field(default_factory=list)
    timeline: list[Paper] = Field(default_factory=list)
    topic_names: list[str] = Field(default_factory=list)
    collaborators: list[dict[str, Any]] = Field(default_factory=list)
    descendant_papers: list[Paper] = Field(default_factory=list)
    referenced_papers: list[Paper] = Field(default_factory=list)
    authored_paper_count: int = 0
    referenced_paper_count: int = 0
    citing_paper_count: int = 0


class ProfessorBriefResponse(ApiModel):
    artifact_id: int | None = None
    created_at: datetime | None = None
    professor: Author
    provider: str
    model: str
    configured: bool
    research_brief: str
    industry_impact: str = ""
    build_on_research: str = ""
    approach_advice: str = ""
    future_directions: list[str] = Field(default_factory=list)
    evidence_ids: list[str] = Field(default_factory=list)
    top_papers: list[Paper] = Field(default_factory=list)
    timeline: list[Paper] = Field(default_factory=list)
    topic_names: list[str] = Field(default_factory=list)
    collaborators: list[dict[str, Any]] = Field(default_factory=list)
    descendant_papers: list[Paper] = Field(default_factory=list)
    referenced_papers: list[Paper] = Field(default_factory=list)
    authored_paper_count: int = 0
    referenced_paper_count: int = 0
    citing_paper_count: int = 0


class GeneratedArtifactSummary(ApiModel):
    id: int
    artifact_type: str
    subject_type: str
    subject_id: str
    created_at: datetime | None = None


class GeneratedArtifactHistoryItem(GeneratedArtifactSummary):
    provider: str
    model: str


class GeneratedArtifactHistoryResponse(ApiModel):
    items: list[GeneratedArtifactHistoryItem] = Field(default_factory=list)


class GeneratedArtifactCatalogItem(GeneratedArtifactSummary):
    provider: str
    model: str
    title: str | None = None


class GeneratedArtifactCatalogResponse(ApiModel):
    items: list[GeneratedArtifactCatalogItem] = Field(default_factory=list)


class DemoBootstrapResponse(ApiModel):
    user_id: str
    provider: str
    model: str
    configured: bool
    featured_topics: list[str] = Field(default_factory=list)
    featured_professors: list[ProfessorSearchResult] = Field(default_factory=list)
